### PR TITLE
feat: Improve cast compatibility tests and docs

### PR DESCRIPTION
**Pull request by andygrove**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion-comet/issues/286

## What changes are included in this PR?

- Improved CometCastSuite to assert that cast tests are enabled (and therefore passing) for casts that we report as compatible. This helps ensure that the docs are in sync with the real status. There is likely more that we can do in the future but this is another step closer.
- Update compatibility reporting based on errors found by this new test enhancement
- Added some notes on some of the current incompatibilities so that they are in the documentation
- Improvements to the documentation formatting
- Fixed a regression from a recent PR where the cast float/double to string tests weren't actually running the test

## Docs Preview

We have a table of compatible casts, followed by:

Screenshot 2024-05-03 at 11 30 15 PM

## How are these changes tested?

Tested docs locally.

### CODE REVIEW

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 1]
```diff
@@ -347,7 +364,7 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       Short.MaxValue.toFloat,
       0.0f) ++
       Range(0, dataSize).map(_ => r.nextFloat())
-    withNulls(values).toDF("a")
+    castTest(withNulls(values).toDF("a"), DataTypes.StringType)
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:367]: Fixing a regression that was introduced in https://github.com/apache/datafusion-comet/pull/346

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 2]
```diff
@@ -401,7 +418,7 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       Double.NegativeInfinity,
       0.0d) ++
       Range(0, dataSize).map(_ => r.nextDouble())
-    withNulls(values).toDF("a")
+    castTest(withNulls(values).toDF("a"), DataTypes.StringType)
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:421]: Fixing a regression that was introduced in https://github.com/apache/datafusion-comet/pull/346

File: spark/src/test/scala/org/apache/comet/CometCastSuite.scala [hunk 3]
```diff
@@ -795,7 +812,7 @@ class CometCastSuite extends CometTestBase with AdaptiveSparkPlanHelper {
       Seq(
         "2024-01-01T12:34:56.123456",
         "2024-01-01T01:00:00Z",
-        "2024-12-31T01:00:00-02:00",
+        "9999-12-31T01:00:00-02:00",
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/CometCastSuite.scala:815]: Just making sure we have a wider range of years being tested

File: spark/src/main/scala/org/apache/comet/GenerateDocs.scala [hunk 4]
```diff
@@ -64,23 +64,36 @@ object GenerateDocs {
     val outputFilename = "docs/source/user-guide/compatibility.md"
     val w = new BufferedOutputStream(new FileOutputStream(outputFilename))
     for (line <- Source.fromFile(templateFilename).getLines()) {
-      if (line.trim == "<!--CAST_TABLE-->") {
-        w.write("| From Type | To Type | Compatible? | Notes |\n".getBytes)
-        w.write("|-|-|-|-|\n".getBytes)
+      if (line.trim == "<!--COMPAT_CAST_TABLE-->") {
```
Comment (andygrove) [inline spark/src/main/scala/org/apache/comet/GenerateDocs.scala:67]: There was previously one table, but now we show separate tables for compatible and incompatible.

File: spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala [hunk 5]
```diff
@@ -253,7 +253,8 @@ class CometExecSuite extends CometTestBase {
     dataTypes.map { subqueryType =>
       withSQLConf(
         CometConf.COMET_EXEC_SHUFFLE_ENABLED.key -> "true",
-        CometConf.COMET_COLUMNAR_SHUFFLE_ENABLED.key -> "true") {
+        CometConf.COMET_COLUMNAR_SHUFFLE_ENABLED.key -> "true",
+        CometConf.COMET_CAST_ALLOW_INCOMPATIBLE.key -> "true") {
```
Comment (andygrove) [inline spark/src/test/scala/org/apache/comet/exec/CometExecSuite.scala:257]: This test needs binary -> string casts

**Comment by andygrove** [comment]
@viirya @snmvaughan This PR incorporates some of the feedback from https://github.com/apache/datafusion-comet/pull/362

**Review by viirya — APPROVED** [review]
Looks good to me. Thanks @andygrove

**Pull request merged**